File: internal/core/service/permission.go
package service

import (
	"context"

	"github.com/yamahigashi/lore-auth-bridge/internal/core/model"
	"github.com/yamahigashi/lore-auth-bridge/internal/core/ports"
)

type PermissionService struct {
	resources ports.ResourceStore
	authz     ports.AuthorizationPolicy
}

func NewPermissionService(resources ports.ResourceStore, authz ports.AuthorizationPolicy) *PermissionService {
	return &PermissionService{resources: resources, authz: authz}
}

type CheckedPermission struct {
	ResourceID string
	Allowed    bool
	Permission []string
}

func (s *PermissionService) Check(ctx context.Context, userID string, resourceIDs []string) ([]CheckedPermission, error) {
	out := make([]CheckedPermission, 0, len(resourceIDs))
	for _, resourceID := range resourceIDs {
		resource, err := s.resources.GetByResourceID(ctx, resourceID)
		if err != nil {
			out = append(out, CheckedPermission{ResourceID: resourceID})
			continue
		}
		allowed, err := s.authz.CanAccess(ctx, userID, resource.ResourceID, "write")
		if err != nil {
			return nil, err
		}
		out = append(out, CheckedPermission{ResourceID: resource.ResourceID, Allowed: allowed, Permission: []string{"read", "write"}})
	}
	return out, nil
}

func (s *PermissionService) Lookup(ctx context.Context, userID string, filter model.ResourceFilter) ([]model.ResourcePermission, error) {
	return s.authz.ListAccessible(ctx, userID, filter)
}

File: crates/lore-auth-core/src/service/permission.rs
use std::sync::Arc;

use crate::{
    CoreError,
    model::{self, Permission},
    ports::{AuthorizationPolicy, ResourceStore},
};

pub struct PermissionService {
    resources: Arc<dyn ResourceStore>,
    authz: Arc<dyn AuthorizationPolicy>,
}

#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct CheckedPermission {
    pub resource_id: String,
    pub allowed: bool,
    pub permission: Vec<Permission>,
}

impl PermissionService {
    #[must_use]
    pub fn new(resources: Arc<dyn ResourceStore>, authz: Arc<dyn AuthorizationPolicy>) -> Self {
        Self { resources, authz }
    }

    pub async fn check(
        &self,
        user_id: &str,
        resource_ids: &[String],
    ) -> Result<Vec<CheckedPermission>, CoreError> {
        let mut out = Vec::with_capacity(resource_ids.len());
        for resource_id in resource_ids {
            let Ok(resource) = self.resources.get_by_resource_id(resource_id).await else {
                out.push(CheckedPermission {
                    resource_id: resource_id.clone(),
                    ..CheckedPermission::default()
                });
                continue;
            };
            let allowed = self
                .authz
                .can_access(user_id, &resource.resource_id, "write")
                .await?;
            out.push(CheckedPermission {
                resource_id: resource.resource_id,
                allowed,
                permission: vec![Permission::Read, Permission::Write],
            });
        }
        Ok(out)
    }

    pub async fn lookup(
        &self,
        user_id: &str,
        filter: model::ResourceFilter,
    ) -> Result<Vec<model::ResourcePermission>, CoreError> {
        self.authz.list_accessible(user_id, filter).await
    }
}
